common/archiver/gcloud/historyArchiver.go

```diff
@@ -0,0 +1,402 @@
+			return errUploadNonRetriable
+		}
+
+		filename := constructHistoryFilenameMultipart(request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, part)
+		if exist, _ := h.gcloudStorage.Exist(ctx, URI, filename); !exist {
+			if err := h.gcloudStorage.Upload(ctx, URI, filename, encodedHistoryPart); err != nil {
+				logger.Error(archiver.ArchiveTransientErrorMsg, tag.ArchivalArchiveFailReason(errWriteFile), tag.Error(err))
+				scope.IncCounter(metrics.HistoryArchiverArchiveTransientErrorCount)
+				return err
+			}
+
+			totalUploadSize = totalUploadSize + int64(binary.Size(encodedHistoryPart))
+		}
+
+		saveHistoryIteratorState(ctx, featureCatalog, historyIterator, part, &progress)
+	}
+
+	scope.AddCounter(metrics.HistoryArchiverTotalUploadSize, totalUploadSize)
+	scope.AddCounter(metrics.HistoryArchiverHistorySize, totalUploadSize)
+	scope.IncCounter(metrics.HistoryArchiverArchiveSuccessCount)
+	return
+}
+
+// Get is used to access an archived history. When context expires method should stop trying to fetch history.
+// The URI identifies the resource from which history should be accessed and it is up to the implementor to interpret this URI.
+// This method should thrift errors - see filestore as an example.
+func (h *historyArchiver) Get(ctx context.Context, URI archiver.URI, request *archiver.GetHistoryRequest) (*archiver.GetHistoryResponse, error) {
+
+	err := h.validateURI(ctx, URI)
+	if err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidURI.Error()}
+	}
+
+	if err := archiver.ValidateGetRequest(request); err != nil {
+		return nil, &shared.BadRequestError{Message: archiver.ErrInvalidGetHistoryRequest.Error()}
+	}
+
+	var token *getHistoryToken
+	if request.NextPageToken != nil {
+		token, err = deserializeGetHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, &shared.BadRequestError{Message: archiver.ErrNextPageTokenCorrupted.Error()}
+		}
+	} else if request.CloseFailoverVersion != nil {
+		_, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *request.CloseFailoverVersion,
+			HighestPart:          *historyhighestPart,
+			CurrentPart:          *historyCurrentPart,
+			NextBatchIdx:         0,
+		}
+	} else {
+		highestVersion, historyhighestPart, historyCurrentPart, err := h.getHighestVersion(ctx, URI, request)
+		if err != nil {
+			return nil, &shared.InternalServiceError{Message: err.Error()}
+		}
+		token = &getHistoryToken{
+			CloseFailoverVersion: *highestVersion,
+			HighestPart:          *historyhighestPart,
+			CurrentPart:          *historyCurrentPart,
+			NextBatchIdx:         0,
+		}
+	}
+
+	response := &archiver.GetHistoryResponse{}
+	numOfEvents := 0
+	numOfBatches := 0
+
+	for partNum := token.CurrentPart; partNum <= token.HighestPart; partNum++ {
```

Comment:
Seems like `token.CurrentPart` is never updated.

Reply:
Done!